File: src/app/modules/appsConfigurations/domain.ts
import { IdWithIdentifier as JsonIdWithIdentifier } from '../../../YamlSettings.interface';

import { Rect } from '../shared/app/Rect';

export enum ApplicationOptions {
  Float = 'float',
  Unmanage = 'unmanage',
  Force = 'force',
  ObjectNameChange = 'object_name_change',
  Layered = 'layered',
  /*BorderOverflow = 'border_overflow',*/ // present in komorebi but deprecated
  TrayAndMultiWindow = 'tray_and_multi_window',
}

export const LabelByAppOption: Record<ApplicationOptions, string> = {
  [ApplicationOptions.Float]: 'Float',
  /*[ApplicationOptions.BorderOverflow]: 'Border Overflow',*/
  [ApplicationOptions.Force]: 'Forced',
  [ApplicationOptions.Layered]: 'Layered',
  [ApplicationOptions.ObjectNameChange]: 'Name Change',
  [ApplicationOptions.TrayAndMultiWindow]: 'MultiWindow',
  [ApplicationOptions.Unmanage]: 'Unmanaged',
};

export enum ApplicationIdentifier {
  Exe = 'Exe',
  Class = 'Class',
  Title = 'Title',
}

export enum MatchingStrategy {
  Legacy = 'Legacy',
  Equals = 'Equals',
  StartsWith = 'StartsWith',
  EndsWith = 'EndsWith',
  Contains = 'Contains',
  Regex = 'Regex',
}

export interface IdWithIdentifier {
  kind: ApplicationIdentifier;
  id: string;
  matching_strategy: MatchingStrategy;
}

type AppConfigurationsOptions = { [K in ApplicationOptions]: boolean };
export interface AppConfiguration extends AppConfigurationsOptions {
  name: string;
  category: string | null;
  workspace: string | null;
  monitor: number | null;
  kind: ApplicationIdentifier;
  identifier: string;
  matchingStrategy: MatchingStrategy;
  invisibleBorders: Rect.plain | null;
}

export interface AppConfigurationExtended extends AppConfiguration {
  key: number;
  isTemplate?: boolean;
  templateName?: string;
  templateDescription?: string;
}

export class AppConfiguration {
  static default(): AppConfiguration {
    return {
      name: 'New App',
      category: null,
      workspace: null,
      monitor: null,
      kind: ApplicationIdentifier.Exe,
      identifier: 'new-app.exe',
      matchingStrategy: MatchingStrategy.Equals,
      [ApplicationOptions.Float]: false,
      /*[ApplicationOptions.BorderOverflow]: false,*/
      [ApplicationOptions.Force]: false,
      [ApplicationOptions.Layered]: false,
      [ApplicationOptions.ObjectNameChange]: false,
      [ApplicationOptions.TrayAndMultiWindow]: false,
      [ApplicationOptions.Unmanage]: false,
      invisibleBorders: null,
    };
  }

  static from(json_identifier: JsonIdWithIdentifier): AppConfiguration {
    return {
      ...AppConfiguration.default(),
      name: json_identifier.id,
      identifier: json_identifier.id,
      kind: json_identifier.kind as ApplicationIdentifier,
      matchingStrategy: json_identifier.matching_strategy as MatchingStrategy ?? MatchingStrategy.Legacy,
    };
  }
}

File: src/background/state/domain.rs
use serde::{Deserialize, Serialize};

// ============== THEMES ==============

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default, rename_all = "camelCase")]
pub struct ThemeCss {
    pub weg: String,
    pub toolbar: String,
    pub wm: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default, rename_all = "camelCase")]
pub struct ThemeInfo {
    pub display_name: String,
    pub author: String,
    pub description: String,
    pub filename: String,
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default, rename_all = "camelCase")]
pub struct Theme {
    pub info: ThemeInfo,
    pub styles: ThemeCss,
}

// ============== PLACEHOLDERS ==============

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default, rename_all = "camelCase")]
pub struct PlaceholderInfo {
    pub display_name: String,
    pub author: String,
    pub description: String,
    pub filename: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default, rename_all = "camelCase")]
pub struct Placeholder {
    pub info: PlaceholderInfo,
    pub left: Vec<serde_yaml::Value>,
    pub center: Vec<serde_yaml::Value>,
    pub right: Vec<serde_yaml::Value>,
}

// ============== WEG ==============

pub type WegItems = serde_yaml::Value;

// ============== SETTINGS BY APP ==============

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
#[serde(rename_all(deserialize = "snake_case"))]
pub enum AppExtraFlag {
    Float,
    Force,
    Unmanage,
    Pinned,
    // only for backwards compatibility
    ObjectNameChange,
    Layered,
    BorderOverflow,
    TrayAndMultiWindow,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub enum AppIdentifierType {
    #[serde(alias = "exe")]
    Exe,
    #[serde(alias = "class")]
    Class,
    #[serde(alias = "title")]
    Title,
    #[serde(alias = "path")]
    Path,
}

#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize)]
pub enum MatchingStrategy {
    #[serde(alias = "equals")]
    Equals,
    #[serde(alias = "startsWith")]
    StartsWith,
    #[serde(alias = "endsWith")]
    EndsWith,
    #[serde(alias = "contains")]
    Contains,
    #[serde(alias = "regex")]
    Regex,
    // only for backwards compatibility
    #[serde(alias = "legacy")]
    Legacy,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct AppIdentifier {
    pub id: String,
    pub kind: AppIdentifierType,
    pub matching_strategy: MatchingStrategy,
    #[serde(default)]
    pub negation: bool,
    #[serde(default)]
    pub and: Vec<AppIdentifier>,
    #[serde(default)]
    pub or: Vec<AppIdentifier>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct AppConfig {
    pub name: String,
    pub category: Option<String>,
    pub bound_monitor_idx: Option<usize>,
    pub bound_workspace_name: Option<String>,
    pub identifier: AppIdentifier,
    #[serde(default)]
    pub options: Vec<AppExtraFlag>,
    #[serde(default)]
    pub is_bundled: bool,
}
